Feat/#22: 백오피스 추가기능 구현

jay1261 opened the pull request:
## 📌 연관된 이슈
<!-- 수정하려는 현재 동작을 설명하거나 관련 문제에 대한 링크를 제공해주세요 -->
> Issue Number : #22 

## PR 종류
해당 PR은 어떤 부분에 대한 변화인가요?
<!-- Please check the one that applies to this PR using "x". -->
- [ ] 버그 수정 `FIX`(Hotfix)
- [ ] 코드 수정 `MODIFY`
- [ ] 기능 추가 `FEAT`
- [ ] 파일 이름 변경 `RENAME`
- [ ] 테스트 코드 추가 `TEST`
- [ ] 코드 스타일 변화 (formatting, local variables) `STYLE`
- [ ] 리팩토링 (no functional changes, no api changes) `REFACTOR`
- [ ] 빌드 관련 내용 추가 및 변화 `BUILD`
- [ ] 문서 내용 추가 및 변화`DOCS`
- [ ] 기타 변경 사항 추 `CHORE`
- [ ] Other... (적어주세요 : )

## 💻 작업 내용
> 이번 PR에서 작업한 내용을 간략히 설명해 주세요 (이미지 첨부 가능)
1. VIP 고객 5명 조회 
<img width="500" alt="VIP고객조회" src="https://github.com/llocr/SpartaEats_Manager/assets/99792672/b8b71521-a5d1-4824-84a2-2fcd48e0950b">

2. 최대 주문 메뉴 5개 조회 
<img width="509" alt="스크린샷 2024-06-23 오후 7 30 09" src="https://github.com/llocr/SpartaEats_Manager/assets/99792672/7615e529-5e2a-42ee-a27b-8e3cc4aa91ff">

3. 날짜별 주문 건수 조회
<img width="421" alt="스크린샷 2024-06-23 오후 7 30 16" src="https://github.com/llocr/SpartaEats_Manager/assets/99792672/39c2c4e7-6b6d-4c11-94eb-cfd787355cc5">


## 해당 PR이 다른 영역에 영향을 미치나요?
- [ ] Yes
- [x] No

<!-- "Yes"를 선택한 경우, 어떤 영향을 끼치는지, 발생 위치 Path를 설명해주세요. -->

## 💬 리뷰 요구사항
> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요

- AnalyticsService 에서 일단 repository들을 가져와서 작업을 하긴 했는데, service로 변경을 해야할지 봐주시면 감사합니다.
- 기능 특성상 조회하는 데이터가 많을 것으로 생각되어서, 데이터들을 불러와서 필터하는 것 보다 DB조회 단계에서 쿼리로 하는게 효율적 일 것 같아서 JPQL 사용해서 조회했습니다.

jay1261 committed:
feat/#22: VIP 고객 리스트 (주문 많이한 순, 5개)

jay1261 committed:
fix/#22: test method 이름 변경

jay1261 committed:
fix/#22: 최대 주문 메뉴 보기 (주문 많은 순 5개)

jay1261 committed:
fix/#22: 날짜별 주문 건수

jay1261 committed:
refactor/#22: 메서드명 수정 및 주석 추가

llocr commented on src/main/java/like/heocholi/spartaeats/repository/OrderMenuRepository.java:
```diff
@@ -1,10 +1,24 @@
+            " WHERE om.menu IN :menuList" +
+            " GROUP BY om.menu" +
+            " ORDER BY SUM(om.count) DESC" +
+            " LIMIT 5")
```
직접 쿼리문 작성하는 것 좋네요!

llocr commented on src/main/java/like/heocholi/spartaeats/service/AnalyticsService.java:
- 이후에 가능하다면 `List<Object[]>` 보다는  아래와 같이 쿼리문이 반환하는 타입을 받을 수 있는 객체를 생성해서 받아보는 것이 좋을 것 같습니다!
다른 사람들이 어떤 객체를 반환하는 지 확인이 가능하고, 런타임에서 타입 오류가 나는 걸 방지할 수 있을 것 같아요 :)
일단은 지금 방식으로 진행해도 될 것 같긴 합니다!
```java
public class VipAndOrderCountDTO {
    private Customer customer;
    private Long orderCount;
}
```
- `System.out.println()`을 사용해서 디버깅 정보를 출력하기보다는 `@Slf4j`를 사용해서 로그를 출력하는 게 어떠실까요?!

llocr commented on src/main/java/like/heocholi/spartaeats/service/AnalyticsService.java:
```diff
@@ -1,10 +1,101 @@
+    public List<VipResponseDto> getVipList(Long storeId, Manager manager) {
+        Store store = storeService.findStore(storeId, manager);
+
+        List<Object[]> vipAndOrderCountList = orderRepository.getVipAndOrderCountList(store).orElseThrow(
+                ()-> new AnalyticsException(ErrorType.NOT_FOUND_ORDER)
+        );
```
이후를 생각한다면, 이런 코드들은 `orderService`로 분리하면 좋긴 할 것 같네요!

llocr reviewed (approved):
주말에 수고하셨어요! 👍
일단 지금 코드 그대로 진행해도 될 것 같긴 합니다!
시간이 되면 리팩토링 진행해보셔도 좋을 것 같습니다 :)

jay1261 committed:
refactor/#22: 리팩토링

1. AnalyticsService에서 repository들 service로 변경
2. Repository에서 반환값 object[]에서 dto로 변경
3. sout을 log로 변경

llocr commented on src/main/java/like/heocholi/spartaeats/repository/OrderRepository.java:
```diff
@@ -16,19 +18,19 @@ public interface OrderRepository extends JpaRepository<Order, Long> {
+			"new like.heocholi.spartaeats.dto.VipResponseDto(o.customer, COUNT(o.customer))"+
 			" FROM Order o" +
 			" WHERE o.store = :store" +
 			" GROUP BY o.customer" +
 			" ORDER BY COUNT(o.customer) DESC" +
 			" LIMIT 5")
-	Optional<List<Object[]>> getVipAndOrderCountList(Store store);
+	Optional<List<VipResponseDto>> getVips(Store store);
 
 
 	@Query(value = "SELECT " +
-			"DATE(o.createdAt), COUNT(o), SUM(o.totalPrice) " +
+			"new like.heocholi.spartaeats.dto.DailySalesResponseDto(DATE(o.createdAt), COUNT(o), SUM(o.totalPrice)) " +
 			"FROM Order o " +
 			"WHERE o.store = :store " +
 			"GROUP BY DATE(o.createdAt)")
-	Optional<List<Object[]>> getDailySales(Store store);
+	Optional<List<DailySalesResponseDto>> getDailySales(Store store);
```
변경해주신 부분 너무 좋습니다! 👍
